python/docs/index.rst

```diff
@@ -1,11 +1,134 @@
+which to receive messages. The receiver may be found at :class:`proton.Receiver`. Note that senders are most commonly
+obtained by using the convenience method :meth:`proton.reactor.Container.create_receiver`.
+
+Message Delivery
+----------------
+
+The process of sending a message is known as **delevery**. Each sent message has a delivry object which tracks the
+status of the message as it is sent from the sender to the receiver. This also includes actions such as settling the
+delivery, a mechanism by which each peer can verify that the message has been properly received and is either accepted
+or rejected. The delivery class may be found at :class:`proton.Delivery`. The delivery object is most commonly obtained
+when a message-related event occurs through the event object. See `Event Handlers`_ below.
+
+Event Handlers
+--------------
+
+A **handler** is a class that handles events associated with the sending and receiving of messages. This includes
+callbacks for events such as the arrival of a new message, error conditions that might arise, and the closing
+of the connection over which messages are sent. An application developer must handle some key events to
+successfully send and receive messages. When an event handler callback is called by the library, an Event object is
+passed to it which contains details of the event, and in some cases, objects associated with the event. For example,
+when a message is received, the event object will have the property ``event.message`` by which the message itself may
+be obtained. See :class:`proton.Event` for more details.
+
+The following are some of the important event callbacks that should be implemented by a developer:
+
+* **on_start()**: This indicates that the event loop in the container has started, and that a new sender and/or
+    receiver may now be created.
+
+To send a message, the following events need to be handled:
+
+* **on_sendable()**: This callback indicates that send credit has now been set by the receiver, and that a message may
+    now be sent.
+* **on_accepted()**: This callback indicates that a message has been received and accepted by the receiving peer.
+
+To receive a message, the following event need to be handled:
+
+* **on_message()**: This callback indicates that a message has been received. The message and its delivery object may
+    be retreived, and if needed, the message can be either accepted or rejected.
+
+Many other events exist for the handling of transactions and other message events and errors, and if present in
+your handler will be called as the corresponding events arise. See the :ref:`tutorial` for examples of handling
+some other events.
+
+Several event handlers are provided which provide default behavior for most events. These may be found in the
+``proton.handlers`` module. The :class:`proton.handlers.MessagingHandler` is the most commonly used handler for
```

Comment:
I think that the API we provide is actually MessagingHandler - your application needs to inherit from that - the other handlers aren't intended for user use (at least not any more) - so specifically we should not be suggesting that anyone uses handlers inherited directly from proton.Handler.

Also my opinion is that none of the other predefined handlers are for the user to use currently (they are actually aggregated inside MessagingHandler anyway) except perhaps the transaction related handlers (which I know little about). This is partly because I want to view then as implementation detail and get rid of them as separate entities in the future (although this may need a major release revision)

Reply:
How about appending the following to this paragraph:

> Developers would typically directly inherit from this handler to provide default
> behavior, and override callbacks as needed to provide specialized behavior they
> may require.